File: IdentityX/testing_framework/register_test.go
package testing

import (
	"GoAuth/internal/apierr"
	"fmt"
	"net/http"
	"strings"
	"testing"
)

func testRegister(t *testing.T, suite *TestSuite) {
	t.Run("Validation", func(t *testing.T) {
		for _, spec := range ValidationTests {
			spec := spec // capture range variable
			t.Run(spec.Name, func(t *testing.T) {
				client := suite.NewClient(t)
				client.POST("/auth/register").
					WithBody(map[string]string{
						"email":    spec.Email,
						"password": spec.Pass,
					}).
					Expect(http.StatusBadRequest).
					HasErrID(apierr.RequestValidationError).
					ValidationError(spec.Errors...)
			})
		}
	})

	t.Run("WeakPasswords", func(t *testing.T) {
		for i, spec := range WeakPasswordTests {
			spec := spec // capture range variable
			i := i       // capture range variable
			t.Run(spec.Name, func(t *testing.T) {
				client := suite.NewClient(t)
				client.POST("/auth/register").
					WithBody(map[string]string{
						"email":    fmt.Sprintf("weak%d@mail.com", i),
						"password": spec.Password,
					}).
					Expect(http.StatusBadRequest).
					HasErrID(apierr.RequestValidationError).
					ValidationError(spec.Errors...)
			})
		}
	})

	t.Run("PasswordTooLong", func(t *testing.T) {
		client := suite.NewClient(t)
		longPass := "A1@" + strings.Repeat("a", 70) // 3 chars + 70 chars = 73 chars > 72
		client.POST("/auth/register").
			WithBody(map[string]string{
				"email":    "longpass@mail.com",
				"password": longPass,
			}).
			Expect(http.StatusBadRequest).
			HasErrID(apierr.RequestValidationError).
			ValidationError("password must be at most 72 characters long")
	})

	// FIXME: Current validation package does not consider trailing spaces as a valid email, fix it and start using the trailing spaces email
	t.Run("EmailNormalization", func(t *testing.T) {
		client := suite.NewClient(t)
		// email := "  MixedCase@Example.Com  "
		email := "MixedCase@Example.Com"
		normalizedEmail := "mixedcase@example.com"
		password := ValidPassword

		// Register with messy email
		client.POST("/auth/register").
			WithBody(map[string]string{
				"email":    email,
				"password": password,
			}).
			Expect(http.StatusCreated).
			HasMessage("Registered user")

		// Login with normalized email
		client.POST("/auth/login").
			WithBody(map[string]string{
				"email":    normalizedEmail,
				"password": password,
			}).
			Expect(http.StatusOK).
			HasMessage("Logged in")

		// Login with messy email again (normalization happens on input)
		client.POST("/auth/login").
			WithBody(map[string]string{
				"email":    email,
				"password": password,
			}).
			Expect(http.StatusOK).
			HasMessage("Logged in")

		// Try to register again with normalized email - should conflict
		client.POST("/auth/register").
			WithBody(map[string]string{
				"email":    normalizedEmail,
				"password": password,
			}).
			Expect(http.StatusConflict).
			HasErrID(apierr.AuthEmailAlreadyUsed).
			HasMessage("email already in use")
	})

	t.Run("Success", func(t *testing.T) {
		client := suite.NewClient(t)
		client.WithCredentials("new@mail.com", ValidPassword).Register()
	})

	t.Run("DuplicateEmail", func(t *testing.T) {
		client := suite.NewClient(t)
		email := "duplicate@mail.com"
		client.WithCredentials(email, ValidPassword).Register()

		client.POST("/auth/register").
			WithBody(map[string]string{
				"email":    email,
				"password": ValidPassword,
			}).
			Expect(http.StatusConflict).
			HasErrID(apierr.AuthEmailAlreadyUsed).
			HasMessage("email already in use")
	})
}

File: IdentityX/testing_framework/register.test.ts
import { describe, test } from "vitest";
import { createClient } from "./helpers/index.js";
import {
    assertErrID,
    assertMessage,
    assertValidationError,
    shouldFail,
} from "./helpers/assert.js";
import {
    ValidPassword,
    ValidationTests,
    WeakPasswordTests,
} from "./fixtures/auth/testdata.js";

// Error IDs - mirror Go errx constants
const ErrRequestValidation = "0_REQ_0004_D";
const ErrEmailAlreadyUsed  = "0_AUTH_0000_D";

// ============================================================================
// REGISTER TESTS
// ============================================================================

describe("Register", () => {

    describe("Validation", () => {
        for (const spec of ValidationTests) {
            test(spec.name, async () => {
                const client = createClient();
                const body = await shouldFail(
                    client.http.post("/auth/register", { email: spec.email, password: spec.pass }),
                    400
                );
                assertErrID(body, ErrRequestValidation);
                assertValidationError(body, ...spec.errors);
            });
        }
    });

    describe("WeakPasswords", () => {
        for (const [i, spec] of WeakPasswordTests.entries()) {
            test(spec.name, async () => {
                const client = createClient();
                const body = await shouldFail(
                    client.http.post("/auth/register", {
                        email: `weak${i}@mail.com`,
                        password: spec.password,
                    }),
                    400
                );
                assertErrID(body, ErrRequestValidation);
                assertValidationError(body, ...spec.errors);
            });
        }
    });

    test("PasswordTooLong", async () => {
        const client = createClient();
        const longPass = "A1@" + "a".repeat(70); // 73 chars > 72 limit
        const body = await shouldFail(
            client.http.post("/auth/register", {
                email: "longpass@mail.com",
                password: longPass,
            }),
            400
        );
        assertErrID(body, ErrRequestValidation);
        assertValidationError(body, "password must be at most 72 characters long");
    });

    // FIXME: Current validation package does not consider trailing spaces as a
    // valid email. Once fixed, switch email to "  MixedCase@Example.Com  ".
    test("EmailNormalization", async () => {
        const client = createClient();
        const email           = "MixedCase@Example.Com";
        const normalizedEmail = "mixedcase@example.com";

        // Register with mixed-case email
        await client.http.post("/auth/register", { email, password: ValidPassword });

        // Login with normalized email
        const loginNorm = await client.http.post("/auth/login", {
            email: normalizedEmail,
            password: ValidPassword,
        });
        assertMessage(loginNorm.data, "Logged in");

        // Login with original mixed-case email (normalization on input)
        const loginOrig = await client.http.post("/auth/login", {
            email,
            password: ValidPassword,
        });
        assertMessage(loginOrig.data, "Logged in");

        // Re-register with normalized email — should conflict
        const body = await shouldFail(
            client.http.post("/auth/register", {
                email: normalizedEmail,
                password: ValidPassword,
            }),
            409
        );
        assertErrID(body, ErrEmailAlreadyUsed);
        assertMessage(body, "email already in use");
    });

    test("Success", async () => {
        const client = createClient();
        await client.withCredentials("new@mail.com", ValidPassword).register();
    });

    test("DuplicateEmail", async () => {
        const client = createClient();
        const email = "duplicate@mail.com";

        await client.withCredentials(email, ValidPassword).register();

        const body = await shouldFail(
            client.http.post("/auth/register", { email, password: ValidPassword }),
            409
        );
        assertErrID(body, ErrEmailAlreadyUsed);
        assertMessage(body, "email already in use");
    });
});